mitxgraders/comparers/linear_comparer.py

```diff
@@ -154,29 +156,30 @@ def check_comparing_zero(comparer_params_evals, student_evals, tolerance):
         """
         student_zero = all([
             is_nearly_zero(x, tolerance, reference=y)
-            for x, y in zip(comparer_params_evals, student_evals)
+            for x, y in zip(student_evals, comparer_params_evals)
         ])
-        expected_zero = all(x == 0.0 for [x] in comparer_params_evals)
+        expected_zero = all(np.all(x == 0.0) for [x] in comparer_params_evals)
         return student_zero or expected_zero
 
-
     def get_valid_modes(self, is_comparing_zero):
         """
         Returns a copy of self.modes, first removing 'proportional' and 'linear'
         when is_comparing_zero is truthy.
         """
-        remove_if_zero = set(['linear', 'proportional'])
-        invalid = lambda mode: is_comparing_zero and mode in remove_if_zero
-        return tuple(mode for mode in self.modes if not invalid(mode) )
+        if is_comparing_zero:
+            return tuple(mode for mode in self.modes
+                         if mode in self.zero_compatible_modes)
+        return self.modes
 
     def __call__(self, comparer_params_evals, student_evals, utils):
         student_evals_norm = np.linalg.norm(student_evals)
 
         # Validate student input shape...only needed for MatrixGrader
         if hasattr(utils, 'validate_shape'):
             # in numpy, scalars have empty tuples as their shapes
-            scalar_expected = isinstance(comparer_params_evals[0], Number)
-            shape = tuple() if scalar_expected else comparer_params_evals[0].shape
+            expected_0 = comparer_params_evals[0][0]
+            scalar_expected = isinstance(expected_0, Number)
+            shape = tuple() if scalar_expected else expected_0.shape
```

Comment:
This isn't a bug that has a corresponding issue in equality_comparer, is it?

Reply:
No. Despite the same argument name, correlated comparers have an extra level of nesting.

In traditional comparers, `comparer_params_evals` is `[param0_eval, param1_eval,...]`.

In correlated comparers, `comparer_params_evals` is `[[param0_eval0, param1_eval0,...], [param_1_eval0, ...], ...]`

I should rename the argument somehow...